dotnet/src/SemanticKernel/Reliability/DefaultHttpRetryHandler.cs

```diff
@@ -0,0 +1,259 @@
+    /// If the request throws an exception that is not a retryable exception, it is not retried.
+    /// If the request returns a response with a retryable error code, it is retried.
+    /// If the request returns a response with a non-retryable error code, it is not retried.
+    /// If the exception contains a RetryAfter header, the request is retried after the specified delay.
+    /// If configured to use exponential backoff, the delay is doubled for each retry.
+    /// </remarks>
+    /// <param name="request">The request.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        int retryCount = 0;
+
+        var start = this._timeProvider.GetCurrentTime();
+        while (true)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            TimeSpan waitFor = default;
+            string reason = string.Empty;
+            HttpResponseMessage? response = null;
+            try
+            {
+                response = await base.SendAsync(request, cancellationToken);
+
+                // If the request does not require a retry then we're done
+                if (!this.ShouldRetry(response.StatusCode))
+                {
+                    return response;
+                }
+
+                // Drain response content to free connections. Need to perform this
+                // before retry attempt and before the TooManyRetries ServiceException.
+                if (response.Content != null)
+                {
+#if NET5_0_OR_GREATER
+                    await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
+#else
+                    await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
+#endif
+                }
+
+                reason = response.StatusCode.ToString();
+
+                if (retryCount >= this._config.MaxRetryCount)
+                {
+                    this._log.LogError(
+                        "Error executing request, max retry count reached. Reason: {0}", reason);
+                    return response;
+                }
+
+                // If the retry delay is longer than the total timeout, then we'll
+                // just return
+                if (!this.HasTimeForRetry(start, retryCount, response, out waitFor))
+                {
+                    this._log.LogError(
+                        "Error executing request, max total retry time reached. Reason: {0}", reason);
+                    return response;
+                }
+            }
+            catch (Exception e) when ((this.ShouldRetry(e) || this.ShouldRetry(e.InnerException)) &&
+                                      retryCount < this._config.MaxRetryCount &&
+                                      this.HasTimeForRetry(start, retryCount, response, out waitFor))
```

Comment:
* e.InnerException might be null
* nit: before checking the exceptions I would check the counter
* it's a lot of code for a `when`
* when the exception is not caught, the exception bubbles up without logging. I think the way this catch all is implemented makes the code a bit hard to read and maintain